src/main/preload/externalAPI.ts

```diff
@@ -0,0 +1,470 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {IpcRendererEvent, contextBridge, ipcRenderer, webFrame} from 'electron';
+
+import {ExternalAPI} from 'types/externalAPI';
+
+import {DesktopAPI} from '@mattermost/desktop-api';
+
+import {
+    NOTIFY_MENTION,
+    IS_UNREAD,
+    UNREAD_RESULT,
+    SESSION_EXPIRED,
+    REACT_APP_INITIALIZED,
+    USER_ACTIVITY_UPDATE,
+    BROWSER_HISTORY_PUSH,
+    APP_LOGGED_IN,
+    APP_LOGGED_OUT,
+    GET_VIEW_INFO_FOR_TEST,
+    DESKTOP_SOURCES_RESULT,
+    VIEW_FINISHED_RESIZING,
+    CALLS_JOIN_CALL,
+    CALLS_JOINED_CALL,
+    CALLS_LEAVE_CALL,
+    DESKTOP_SOURCES_MODAL_REQUEST,
+    CALLS_WIDGET_SHARE_SCREEN,
+    CALLS_ERROR,
+    CALLS_JOIN_REQUEST,
+    GET_IS_DEV_MODE,
+    TOGGLE_SECURE_INPUT,
+    GET_APP_INFO,
+    REQUEST_BROWSER_HISTORY_STATUS,
+    BROWSER_HISTORY_STATUS_UPDATED,
+    NOTIFICATION_CLICKED,
+    CALLS_WIDGET_RESIZE,
+    CALLS_WIDGET_CHANNEL_LINK_CLICK,
+    CALLS_LINK_CLICK,
+    CALLS_POPOUT_FOCUS,
+    GET_DESKTOP_SOURCES,
+    UNREADS_AND_MENTIONS,
+    LEGACY_OFF,
+} from 'common/communication';
+
+const createListener: ExternalAPI['createListener'] = (channel: string, listener: (...args: never[]) => void) => {
```

Comment:
This is really nicely handled. Something to be extremely careful with is accidentally returning privileged objects to the webapp context. `IpcRenderer` is one such object and this covers that case well.

Reply:
FWIW, I believe Electron will not send anything that isn't serializable, so eg. we can't send a function callback by accident.

Reply:
It'll serialize all objects and proxy all functions. Prototypes chains will be ignored though.